File: src/dynamixel/driver.py
"""Dynamixel SDK driver for ViperX-300s arm."""

import threading
from typing import Dict, List, Optional

from dynamixel_sdk import (
    PortHandler,
    PacketHandler,
    COMM_SUCCESS,
)

from .constants import (
    PROTOCOL_VERSION,
    ADDR_TORQUE_ENABLE,
    ADDR_GOAL_POSITION,
    ADDR_PRESENT_POSITION,
    ADDR_MOVING,
    ADDR_PROFILE_VELOCITY,
    ADDR_PROFILE_ACCELERATION,
    ALL_ARM_MOTOR_IDS,
    PRIMARY_MOTOR_IDS,
    GRIPPER_MOTOR_ID,
    VIPERX_300S_JOINTS,
    NUM_JOINTS,
)
from .conversions import ticks_to_radians, radians_to_ticks


class DynamixelError(Exception):
    """Base exception for Dynamixel communication errors."""

    pass


# Module-level singleton storage for drivers by port
_driver_instances: Dict[str, "DynamixelDriver"] = {}
_driver_lock = threading.Lock()


def get_driver(port: str, baud_rate: int) -> "DynamixelDriver":
    """Get or create a singleton driver instance for a port.

    This ensures only one driver exists per serial port, preventing
    conflicts between arm and gripper components.
    """
    global _driver_instances
    with _driver_lock:
        if port not in _driver_instances:
            driver = DynamixelDriver(port, baud_rate)
            driver.open()
            _driver_instances[port] = driver
        return _driver_instances[port]


def close_driver(port: str) -> None:
    """Close and remove a driver instance."""
    global _driver_instances
    with _driver_lock:
        if port in _driver_instances:
            try:
                _driver_instances[port].close()
            except Exception:
                pass
            del _driver_instances[port]


class DynamixelDriver:
    """Driver for communicating with Dynamixel motors on ViperX-300s.

    This driver handles:
    - Serial port management
    - Synchronized reading/writing for multiple motors
    - Dual-motor joint handling (shoulder, elbow)
    - Thread-safe operations
    """

    def __init__(self, port: str, baud_rate: int):
        """Initialize the Dynamixel driver.

        Args:
            port: Serial port path (e.g., "/dev/ttyUSB0")
            baud_rate: Communication baud rate (typically 1000000)
        """
        self._port_name = port
        self._baud_rate = baud_rate
        self._port_handler: Optional[PortHandler] = None
        self._packet_handler: Optional[PacketHandler] = None
        self._lock = threading.RLock()  # Use RLock for reentrant locking
        self._is_open = False

    @property
    def is_open(self) -> bool:
        """Check if the driver connection is open."""
        return self._is_open

    def open(self) -> None:
        """Open connection to motors.

        Raises:
            DynamixelError: If port cannot be opened or configured
        """
        with self._lock:
            if self._is_open:
                return

            self._port_handler = PortHandler(self._port_name)
            self._packet_handler = PacketHandler(PROTOCOL_VERSION)

            if not self._port_handler.openPort():
                raise DynamixelError(f"Failed to open port {self._port_name}")

            if not self._port_handler.setBaudRate(self._baud_rate):
                self._port_handler.closePort()
                raise DynamixelError(f"Failed to set baud rate {self._baud_rate}")

            self._is_open = True

    def close(self) -> None:
        """Close connection to motors."""
        with self._lock:
            if not self._is_open:
                return

            if self._port_handler:
                self._port_handler.closePort()

            self._is_open = False

    def _check_open(self) -> None:
        """Verify driver is open, raise if not."""
        if not self._is_open:
            raise DynamixelError("Driver not open")

    def enable_torque(self, motor_ids: Optional[List[int]] = None) -> None:
        """Enable torque on specified motors.

        Args:
            motor_ids: List of motor IDs, or None for all arm motors
        """
        if motor_ids is None:
            motor_ids = ALL_ARM_MOTOR_IDS

        with self._lock:
            self._check_open()
            for motor_id in motor_ids:
                result, error = self._packet_handler.write1ByteTxRx(
                    self._port_handler, motor_id, ADDR_TORQUE_ENABLE, 1
                )
                if result != COMM_SUCCESS:
                    raise DynamixelError(
                        f"Failed to enable torque on motor {motor_id}: "
                        f"{self._packet_handler.getTxRxResult(result)}"
                    )

    def disable_torque(self, motor_ids: Optional[List[int]] = None) -> None:
        """Disable torque on specified motors.

        Args:
            motor_ids: List of motor IDs, or None for all arm motors
        """
        if motor_ids is None:
            motor_ids = ALL_ARM_MOTOR_IDS

        with self._lock:
            self._check_open()
            for motor_id in motor_ids:
                result, error = self._packet_handler.write1ByteTxRx(
                    self._port_handler, motor_id, ADDR_TORQUE_ENABLE, 0
                )
                if result != COMM_SUCCESS:
                    raise DynamixelError(
                        f"Failed to disable torque on motor {motor_id}: "
                        f"{self._packet_handler.getTxRxResult(result)}"
                    )

    def read_joint_positions(self) -> List[float]:
        """Read current positions of all joints.

        Returns:
            List of 6 joint positions in radians

        Raises:
            DynamixelError: If communication fails
        """
        with self._lock:
            self._check_open()

            # Read positions individually for reliability
            # read4ByteTxRx returns (data, result, error)
            positions = []
            for joint_idx in range(NUM_JOINTS):
                config = VIPERX_300S_JOINTS[joint_idx]
                primary_id = config.motor_ids[0]

                data, result, error = self._packet_handler.read4ByteTxRx(
                    self._port_handler, primary_id, ADDR_PRESENT_POSITION
                )
                if result != COMM_SUCCESS or error != 0:
                    raise DynamixelError(
                        f"Failed to read position for motor {primary_id}: "
                        f"{self._packet_handler.getTxRxResult(result)}"
                    )

                positions.append(ticks_to_radians(data))

            return positions

    def write_joint_positions(self, positions: List[float]) -> None:
        """Write goal positions to all joints.

        For dual-motor joints (shoulder, elbow), both motors receive
        the same position command to stay synchronized.

        Args:
            positions: List of 6 target positions in radians

        Raises:
            ValueError: If wrong number of positions provided
            DynamixelError: If communication fails
        """
        if len(positions) != NUM_JOINTS:
            raise ValueError(f"Expected {NUM_JOINTS} positions, got {len(positions)}")

        with self._lock:
            self._check_open()

            # Write positions individually for reliability
            for joint_idx, position in enumerate(positions):
                config = VIPERX_300S_JOINTS[joint_idx]
                ticks = radians_to_ticks(position)

                # Write to all motors for this joint (handles dual-motor joints)
                for motor_id in config.motor_ids:
                    result, error = self._packet_handler.write4ByteTxRx(
                        self._port_handler, motor_id, ADDR_GOAL_POSITION, ticks
                    )
                    if result != COMM_SUCCESS:
                        raise DynamixelError(
                            f"Failed to write position for motor {motor_id}: "
                            f"{self._packet_handler.getTxRxResult(result)}"
                        )

    def is_moving(self) -> bool:
        """Check if any motor is currently moving.

        Returns:
            True if any motor is moving, False otherwise
        """
        with self._lock:
            self._check_open()

            # Read moving status from each motor individually
            # read1ByteTxRx returns (data, result, error)
            for motor_id in PRIMARY_MOTOR_IDS:
                data, result, error = self._packet_handler.read1ByteTxRx(
                    self._port_handler, motor_id, ADDR_MOVING
                )
                if result != COMM_SUCCESS or error != 0:
                    # Skip motors that fail to respond
                    continue
                if data:
                    return True

            return False

    def set_profile_velocity(
        self, velocity: int, motor_ids: Optional[List[int]] = None
    ) -> None:
        """Set profile velocity for smooth motion.

        Args:
            velocity: Velocity value (0 = no limit, higher = faster)
            motor_ids: List of motor IDs, or None for all arm motors
        """
        if motor_ids is None:
            motor_ids = ALL_ARM_MOTOR_IDS

        with self._lock:
            self._check_open()
            for motor_id in motor_ids:
                result, error = self._packet_handler.write4ByteTxRx(
                    self._port_handler, motor_id, ADDR_PROFILE_VELOCITY, velocity
                )
                if result != COMM_SUCCESS:
                    raise DynamixelError(
                        f"Failed to set velocity on motor {motor_id}"
                    )

    def set_profile_acceleration(
        self, acceleration: int, motor_ids: Optional[List[int]] = None
    ) -> None:
        """Set profile acceleration for smooth motion.

        Args:
            acceleration: Acceleration value (0-32767)
            motor_ids: List of motor IDs, or None for all arm motors
        """
        if motor_ids is None:
            motor_ids = ALL_ARM_MOTOR_IDS

        with self._lock:
            self._check_open()
            for motor_id in motor_ids:
                result, error = self._packet_handler.write4ByteTxRx(
                    self._port_handler, motor_id, ADDR_PROFILE_ACCELERATION, acceleration
                )
                if result != COMM_SUCCESS:
                    raise DynamixelError(
                        f"Failed to set acceleration on motor {motor_id}"
                    )

    # Gripper-specific methods

    def enable_gripper_torque(self) -> None:
        """Enable torque on the gripper motor."""
        self.enable_torque([GRIPPER_MOTOR_ID])

    def disable_gripper_torque(self) -> None:
        """Disable torque on the gripper motor."""
        self.disable_torque([GRIPPER_MOTOR_ID])

    def read_gripper_position(self) -> float:
        """Read current gripper position.

        Returns:
            Gripper position in radians
        """
        with self._lock:
            self._check_open()

            # read4ByteTxRx returns (data, result, error)
            data, result, error = self._packet_handler.read4ByteTxRx(
                self._port_handler, GRIPPER_MOTOR_ID, ADDR_PRESENT_POSITION
            )
            if result != COMM_SUCCESS or error != 0:
                raise DynamixelError(
                    f"Failed to read gripper position: "
                    f"{self._packet_handler.getRxPacketError(error)}"
                )

            return ticks_to_radians(data)

    def write_gripper_position(self, position: float) -> None:
        """Write goal position to gripper.

        Args:
            position: Target position in radians
        """
        with self._lock:
            self._check_open()

            ticks = radians_to_ticks(position)
            result, error = self._packet_handler.write4ByteTxRx(
                self._port_handler, GRIPPER_MOTOR_ID, ADDR_GOAL_POSITION, ticks
            )
            if result != COMM_SUCCESS:
                raise DynamixelError(
                    f"Failed to write gripper position: "
                    f"{self._packet_handler.getTxRxResult(result)}"
                )

    def is_gripper_moving(self) -> bool:
        """Check if gripper is currently moving.

        Returns:
            True if gripper is moving, False otherwise
        """
        with self._lock:
            self._check_open()

            # read1ByteTxRx returns (data, result, error)
            data, result, error = self._packet_handler.read1ByteTxRx(
                self._port_handler, GRIPPER_MOTOR_ID, ADDR_MOVING
            )
            if result != COMM_SUCCESS or error != 0:
                raise DynamixelError(
                    f"Failed to read gripper moving status: "
                    f"{self._packet_handler.getRxPacketError(error)}"
                )

            return bool(data)

File: dynamixel/driver.go
package dynamixel

import (
	"errors"
	"fmt"
	"strings"
	"sync"
	"time"

	protocol "github.com/haguro/go-dxl/protocol/v2"
	"go.bug.st/serial"
)

// ErrNotOpen is returned when operations are attempted on a closed driver.
var ErrNotOpen = errors.New("driver not open")

// Driver provides thread-safe communication with Dynamixel motors.
type Driver struct {
	port    serial.Port
	handler *protocol.Handler
	mu      sync.Mutex
	isOpen  bool
}

// NewDriver creates and opens a new Dynamixel driver.
func NewDriver(portName string, baudRate int) (*Driver, error) {
	mode := &serial.Mode{
		BaudRate: baudRate,
		DataBits: 8,
		Parity:   serial.NoParity,
		StopBits: serial.OneStopBit,
	}

	port, err := serial.Open(portName, mode)
	if err != nil {
		return nil, fmt.Errorf("failed to open serial port %s: %w", portName, err)
	}

	// Set read timeout
	if err := port.SetReadTimeout(100 * time.Millisecond); err != nil {
		port.Close()
		return nil, fmt.Errorf("failed to set read timeout: %w", err)
	}

	handler := protocol.NewHandler(port, 100*time.Millisecond)

	return &Driver{
		port:    port,
		handler: handler,
		isOpen:  true,
	}, nil
}

// Close closes the driver and releases resources.
func (d *Driver) Close() error {
	d.mu.Lock()
	defer d.mu.Unlock()

	if !d.isOpen {
		return nil
	}

	d.isOpen = false
	return d.port.Close()
}

// checkOpen verifies the driver is open.
func (d *Driver) checkOpen() error {
	if !d.isOpen {
		return ErrNotOpen
	}
	return nil
}

// EnableTorque enables torque on the specified motors.
func (d *Driver) EnableTorque(motorIDs []int) error {
	d.mu.Lock()
	defer d.mu.Unlock()

	if err := d.checkOpen(); err != nil {
		return err
	}

	for _, id := range motorIDs {
		if err := d.handler.Write(byte(id), AddrTorqueEnable, 1); err != nil {
			// Ignore hardware errors (motor may have stale error flag)
			if isHardwareError(err) {
				continue
			}
			return fmt.Errorf("failed to enable torque on motor %d: %w", id, err)
		}
	}
	return nil
}

// DisableTorque disables torque on the specified motors.
func (d *Driver) DisableTorque(motorIDs []int) error {
	d.mu.Lock()
	defer d.mu.Unlock()

	if err := d.checkOpen(); err != nil {
		return err
	}

	for _, id := range motorIDs {
		if err := d.handler.Write(byte(id), AddrTorqueEnable, 0); err != nil {
			// Ignore hardware errors
			if isHardwareError(err) {
				continue
			}
			return fmt.Errorf("failed to disable torque on motor %d: %w", id, err)
		}
	}
	return nil
}

// ReadJointPositions reads current positions of all arm joints.
// Returns positions in radians.
func (d *Driver) ReadJointPositions() ([]float64, error) {
	d.mu.Lock()
	defer d.mu.Unlock()

	if err := d.checkOpen(); err != nil {
		return nil, err
	}

	positions := make([]float64, NumJoints)

	// Read from primary motor of each joint
	for i, config := range ViperX300SJoints {
		primaryID := config.MotorIDs[0]
		data, err := d.handler.Read(byte(primaryID), AddrPresentPosition, 4)
		if err != nil {
			return nil, fmt.Errorf("failed to read position from motor %d: %w", primaryID, err)
		}

		ticks := int(BytesToInt32(data))
		positions[i] = TicksToRadians(ticks)
	}

	return positions, nil
}

// WriteJointPositions writes goal positions to all arm joints.
// Positions are in radians.
// Note: Hardware errors (like data limit errors) are logged but not returned,
// matching the Python SDK behavior.
func (d *Driver) WriteJointPositions(positions []float64) error {
	if len(positions) != NumJoints {
		return fmt.Errorf("expected %d positions, got %d", NumJoints, len(positions))
	}

	d.mu.Lock()
	defer d.mu.Unlock()

	if err := d.checkOpen(); err != nil {
		return err
	}

	// Write to each joint's motors
	for i, pos := range positions {
		config := ViperX300SJoints[i]
		ticks := RadiansToTicks(pos)
		data := Int32ToBytes(int32(ticks))

		// Write to all motors for this joint (handles dual-motor joints)
		for _, motorID := range config.MotorIDs {
			if err := d.handler.Write(byte(motorID), AddrGoalPosition, data...); err != nil {
				// Log hardware errors but continue (matches Python SDK behavior)
				// Hardware errors like "data limit error" happen when position
				// exceeds motor's EEPROM limits, but the motor will still move
				// to the closest valid position
				if isHardwareError(err) {
					// Silently ignore hardware errors like Python does
					continue
				}
				return fmt.Errorf("failed to write position to motor %d: %w", motorID, err)
			}
		}
	}

	return nil
}

// isHardwareError checks if the error is a Dynamixel hardware error
// (as opposed to a communication error)
func isHardwareError(err error) bool {
	if err == nil {
		return false
	}
	errStr := err.Error()
	// Hardware errors from Dynamixel contain these phrases
	return strings.Contains(errStr, "data limit error") ||
		strings.Contains(errStr, "processing error") ||
		strings.Contains(errStr, "hardware error") ||
		strings.Contains(errStr, "overload error") ||
		strings.Contains(errStr, "overheating error")
}

// IsMoving checks if any arm motor is currently moving.
func (d *Driver) IsMoving() (bool, error) {
	d.mu.Lock()
	defer d.mu.Unlock()

	if err := d.checkOpen(); err != nil {
		return false, err
	}

	for _, id := range PrimaryMotorIDs {
		data, err := d.handler.Read(byte(id), AddrMoving, 1)
		if err != nil {
			return false, fmt.Errorf("failed to read moving status from motor %d: %w", id, err)
		}
		if len(data) > 0 && data[0] != 0 {
			return true, nil
		}
	}

	return false, nil
}

// SetProfileVelocity sets the profile velocity for smooth motion.
func (d *Driver) SetProfileVelocity(velocity int, motorIDs []int) error {
	d.mu.Lock()
	defer d.mu.Unlock()

	if err := d.checkOpen(); err != nil {
		return err
	}

	data := Int32ToBytes(int32(velocity))
	for _, id := range motorIDs {
		if err := d.handler.Write(byte(id), AddrProfileVelocity, data...); err != nil {
			return fmt.Errorf("failed to set velocity on motor %d: %w", id, err)
		}
	}

	return nil
}

// SetProfileAcceleration sets the profile acceleration for smooth motion.
func (d *Driver) SetProfileAcceleration(acceleration int, motorIDs []int) error {
	d.mu.Lock()
	defer d.mu.Unlock()

	if err := d.checkOpen(); err != nil {
		return err
	}

	data := Int32ToBytes(int32(acceleration))
	for _, id := range motorIDs {
		if err := d.handler.Write(byte(id), AddrProfileAcceleration, data...); err != nil {
			return fmt.Errorf("failed to set acceleration on motor %d: %w", id, err)
		}
	}

	return nil
}

// Reboot reboots a motor to clear hardware errors.
func (d *Driver) Reboot(motorID int) error {
	d.mu.Lock()
	defer d.mu.Unlock()

	if err := d.checkOpen(); err != nil {
		return err
	}

	if err := d.handler.Reboot(byte(motorID)); err != nil {
		return fmt.Errorf("failed to reboot motor %d: %w", motorID, err)
	}

	// Give the motor time to reboot
	time.Sleep(500 * time.Millisecond)
	return nil
}

// ReadOperatingMode reads the operating mode of a motor.
func (d *Driver) ReadOperatingMode(motorID int) (int, error) {
	d.mu.Lock()
	defer d.mu.Unlock()

	if err := d.checkOpen(); err != nil {
		return 0, err
	}

	data, err := d.handler.Read(byte(motorID), AddrOperatingMode, 1)
	if err != nil {
		return 0, fmt.Errorf("failed to read operating mode from motor %d: %w", motorID, err)
	}

	if len(data) == 0 {
		return 0, fmt.Errorf("no data returned for motor %d", motorID)
	}

	return int(data[0]), nil
}

// SetOperatingMode sets the operating mode of a motor.
// Note: Torque must be disabled before changing operating mode.
func (d *Driver) SetOperatingMode(motorID int, mode int) error {
	d.mu.Lock()
	defer d.mu.Unlock()

	if err := d.checkOpen(); err != nil {
		return err
	}

	if err := d.handler.Write(byte(motorID), AddrOperatingMode, byte(mode)); err != nil {
		return fmt.Errorf("failed to set operating mode on motor %d: %w", motorID, err)
	}

	return nil
}

// Gripper methods

// EnableGripperTorque enables torque on the gripper motor.
func (d *Driver) EnableGripperTorque() error {
	return d.EnableTorque([]int{GripperMotorID})
}

// DisableGripperTorque disables torque on the gripper motor.
func (d *Driver) DisableGripperTorque() error {
	return d.DisableTorque([]int{GripperMotorID})
}

// ReadGripperPosition reads the current gripper position in radians.
func (d *Driver) ReadGripperPosition() (float64, error) {
	d.mu.Lock()
	defer d.mu.Unlock()

	if err := d.checkOpen(); err != nil {
		return 0, err
	}

	data, err := d.handler.Read(byte(GripperMotorID), AddrPresentPosition, 4)
	if err != nil {
		return 0, fmt.Errorf("failed to read gripper position: %w", err)
	}

	ticks := int(BytesToInt32(data))
	return TicksToRadians(ticks), nil
}

// WriteGripperPosition writes the goal position to the gripper in radians.
func (d *Driver) WriteGripperPosition(position float64) error {
	d.mu.Lock()
	defer d.mu.Unlock()

	if err := d.checkOpen(); err != nil {
		return err
	}

	ticks := RadiansToTicks(position)
	data := Int32ToBytes(int32(ticks))

	if err := d.handler.Write(byte(GripperMotorID), AddrGoalPosition, data...); err != nil {
		return fmt.Errorf("failed to write gripper position: %w", err)
	}

	return nil
}

// IsGripperMoving checks if the gripper motor is currently moving.
func (d *Driver) IsGripperMoving() (bool, error) {
	d.mu.Lock()
	defer d.mu.Unlock()

	if err := d.checkOpen(); err != nil {
		return false, err
	}

	data, err := d.handler.Read(byte(GripperMotorID), AddrMoving, 1)
	if err != nil {
		return false, fmt.Errorf("failed to read gripper moving status: %w", err)
	}

	return len(data) > 0 && data[0] != 0, nil
}

// Singleton driver management for sharing between arm and gripper components.

var (
	driverInstance *Driver
	driverMu       sync.Mutex
	driverRefCount int
)

// GetDriver returns a shared driver instance for the given port.
// Multiple calls with the same port will return the same driver.
func GetDriver(portName string, baudRate int) (*Driver, error) {
	fmt.Printf("[DEBUG] GetDriver: acquiring lock for port %s\n", portName)
	driverMu.Lock()
	defer driverMu.Unlock()
	fmt.Printf("[DEBUG] GetDriver: lock acquired, driverInstance=%v, refCount=%d\n", driverInstance != nil, driverRefCount)

	if driverInstance != nil {
		driverRefCount++
		fmt.Printf("[DEBUG] GetDriver: reusing existing driver, new refCount=%d\n", driverRefCount)
		return driverInstance, nil
	}

	fmt.Printf("[DEBUG] GetDriver: creating new driver for %s at %d baud\n", portName, baudRate)
	driver, err := NewDriver(portName, baudRate)
	if err != nil {
		fmt.Printf("[DEBUG] GetDriver: NewDriver failed: %v\n", err)
		return nil, err
	}

	driverInstance = driver
	driverRefCount = 1
	fmt.Printf("[DEBUG] GetDriver: new driver created successfully\n")
	return driver, nil
}

// ReleaseDriver decrements the reference count and closes the driver when no longer in use.
func ReleaseDriver() {
	fmt.Printf("[DEBUG] ReleaseDriver: acquiring lock\n")
	driverMu.Lock()
	defer driverMu.Unlock()
	fmt.Printf("[DEBUG] ReleaseDriver: lock acquired, driverInstance=%v, refCount=%d\n", driverInstance != nil, driverRefCount)

	if driverInstance == nil {
		fmt.Printf("[DEBUG] ReleaseDriver: no driver instance to release\n")
		return
	}

	driverRefCount--
	fmt.Printf("[DEBUG] ReleaseDriver: decremented refCount to %d\n", driverRefCount)
	if driverRefCount <= 0 {
		fmt.Printf("[DEBUG] ReleaseDriver: closing driver\n")
		driverInstance.Close()
		driverInstance = nil
		driverRefCount = 0
		fmt.Printf("[DEBUG] ReleaseDriver: driver closed and cleared\n")
	}
}
